File: core/src/recognition/phash.ts
// 8x8 DCT 感知哈希。算法必须与 pipeline/src/ad_pipeline/phash.py 完全一致。

const EPS = 1e-6; // 中位数比较的浮点临界带;必须与 phash.py 的 EPS 同值

function dct1d(vec: number[]): number[] {
  const n = vec.length;
  const out = new Array<number>(n);
  for (let k = 0; k < n; k++) {
    let s = 0;
    for (let i = 0; i < n; i++) {
      s += vec[i] * Math.cos((Math.PI * (2 * i + 1) * k) / (2 * n));
    }
    out[k] = s;
  }
  return out;
}

function dct2d(mat: number[][]): number[][] {
  const rows = mat.map(dct1d);
  const n = rows.length;
  const cols: number[][] = [];
  for (let k = 0; k < rows[0].length; k++) {
    cols.push(dct1d(rows.map((_, i) => rows[i][k])));
  }
  return Array.from({ length: n }, (_, i) =>
    Array.from({ length: cols.length }, (_, k) => cols[k][i]));
}

export function phashFromGray(gray32: number[][]): string {
  const dct = dct2d(gray32.map((r) => r.map(Number)));
  const low: number[] = [];
  for (let r = 0; r < 8; r++) for (let c = 0; c < 8; c++) low.push(dct[r][c]);
  const rest = low.filter((_, i) => i !== 0).slice().sort((a, b) => a - b);
  const m = rest.length;
  const median = m % 2 ? rest[(m - 1) / 2] : (rest[m / 2 - 1] + rest[m / 2]) / 2;
  let bits = 0n;
  for (const v of low) bits = (bits << 1n) | (v > median + EPS ? 1n : 0n);
  return bits.toString(16).padStart(16, "0");
}

export function hamming(a: string, b: string): number {
  let x = BigInt("0x" + a) ^ BigInt("0x" + b);
  let count = 0;
  while (x > 0n) { count += Number(x & 1n); x >>= 1n; }
  return count;
}

File: pipeline/src/ad_pipeline/phash.py
"""8x8 DCT 感知哈希(权威实现,TS 端必须位对齐)。无第三方依赖。

接收已是 32x32 的灰度图(缩放由 resize.resize_gray_to_32 负责)。
"""
import math

EPS = 1e-6  # 中位数比较的浮点临界带;TS 端必须用同一常量


def _dct_1d(vec: list[float]) -> list[float]:
    n = len(vec)
    out = []
    for k in range(n):
        s = 0.0
        for i in range(n):
            s += vec[i] * math.cos(math.pi * (2 * i + 1) * k / (2 * n))
        out.append(s)
    return out


def _dct_2d(mat: list[list[float]]) -> list[list[float]]:
    rows = [_dct_1d(r) for r in mat]
    cols = []
    n = len(rows)
    for k in range(len(rows[0])):
        col = _dct_1d([rows[i][k] for i in range(n)])
        cols.append(col)
    # cols[k][i] = 第 k 列的 DCT;转回 [i][k]
    return [[cols[k][i] for k in range(len(cols))] for i in range(n)]


def phash_from_gray(gray32: list[list[int]]) -> str:
    assert len(gray32) == 32 and len(gray32[0]) == 32
    dct = _dct_2d([[float(v) for v in row] for row in gray32])
    low = [dct[r][c] for r in range(8) for c in range(8)]
    rest = [v for i, v in enumerate(low) if i != 0]  # 排除 DC [0,0]
    rest_sorted = sorted(rest)
    m = len(rest_sorted)
    median = (rest_sorted[m // 2] if m % 2 else
              (rest_sorted[m // 2 - 1] + rest_sorted[m // 2]) / 2)
    bits = 0
    for v in low:
        bits = (bits << 1) | (1 if v > median + EPS else 0)
    return f"{bits:016x}"


def hamming(a: str, b: str) -> int:
    return bin(int(a, 16) ^ int(b, 16)).count("1")
